Add doc loader

hackerdave opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes #

## Change
<!-- Please describe the changes you made. -->
Add doc loader classes for calling the dbms_vector package to convert a rich document to text, chunk, produce vector embeddings, and generate summary.

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [x] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

hackerdave committed:
Add doc loader

psilberk commented on langchain4j-oracle/src/test/java/dev/langchain4j/data/document/loader/OracleDocumentLoaderTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(OracleDocumentLoaderTest.class);
+
+    Dotenv dotenv;
```
I think we should use System.getEnv, just like in the other tests (look at CommonTestOperations.java).

psilberk commented on langchain4j-oracle/src/test/java/dev/langchain4j/data/document/loader/OracleDocumentLoaderTest.java:
```diff
@@ -0,0 +1,89 @@
+        dotenv = Dotenv.configure().load();
+
+        try {
+            Connection conn = DriverManager.getConnection(
```
We should create a PoolDataSource and use it to get connections. Please refer to CommonTestsOperations.java too.

psilberk commented on langchain4j-oracle/src/test/java/dev/langchain4j/data/document/loader/OracleDocumentLoaderTest.java:
```diff
@@ -0,0 +1,89 @@
+            Connection conn = DriverManager.getConnection(
+                    dotenv.get("ORACLE_JDBC_URL"), dotenv.get("ORACLE_JDBC_USER"), dotenv.get("ORACLE_JDBC_PASSWORD"));
+            loader = new OracleDocumentLoader(conn);
+        } catch (SQLException ex) {
```
Instead of all these catchs that might swallow some of the details we can just declare them (void setUp() throws SQLException). Then you don't need to catch the checked exception.

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -0,0 +1,164 @@
+        JsonNode tableNode = rootNode.path("tablename");
+        JsonNode colNode = rootNode.path("colname");
+
+        if (fileNode.textValue() != null) {
```
When I tried with no settings in DEMO_DS_PDF_FILE, the test got in this code path? Was that the intention? The json when this value is not set gets {"file", "null"} which is different than just null.

hackerdave replied:
Good catch, this should throw
java.security.InvalidParameterException: Missing file, dir, or table in preference

Updated to also check for "null", e.g.,
        if (fileNode.textValue() != null && !fileNode.textValue().equals("null")) {

hackerdave committed:
Rebase and other fixes

Spotless refactor
fix jackson dependency warning
check for null string
setup now throws SQLException
Cohere model update

hackerdave committed:
Fix jackson dependency warning

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/splitter/oracle/Chunk.java:
```diff
@@ -0,0 +1,13 @@
+
+public class Chunk {
+
+    public int chunk_id;
```
Is there a reason not to use standard pojo conventions in here: https://docs.oracle.com/cd/E19316-01/819-3669/bnais/index.html (basically, setting the attributes private and add getter and setters)

hackerdave replied:
Both the Chunk and Embedding class are used for JSON serialization and deserialization and aren't expected to be used by a user directly. We can certainly change to a POJO.

hackerdave committed:
Change chunk and embed to use POJO

hackerdave committed:
Change to PoolDataSource

hackerdave committed:
Switch from dotenv to System.getenv

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -0,0 +1,163 @@
+
+            try (ResultSet rs = stmt.executeQuery()) {
+                while (rs.next()) {
+                    String text = rs.getString("text");
```
I think we could use some constants for these names (it gets repeated in the code).

hackerdave replied:
Added the constants COLUMN_TEXT and COLUMN_METADATA

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -0,0 +1,163 @@
+        List<Document> documents = new ArrayList<>();
+
+        String query = String.format(
+                "select dbms_vector_chain.utl_to_text(t.%s, json(?)) text, dbms_vector_chain.utl_to_text(t.%s, json('{\"plaintext\": \"false\"}')) metadata from %s.%s t",
```
I think we should add a thorough comment explaning what this SQL is doing and its parameters. We should also format the string so it fits on 80 characters

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -0,0 +1,163 @@
+        List<Document> documents = new ArrayList<>();
+
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode rootNode = mapper.readTree(pref);
```
Can we do here something similar to what you did in Chunk? Using annotations and encapsulating the errors.

hackerdave replied:
Added a LoaderPreference class

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/splitter/oracle/OracleDocumentSplitter.java:
```diff
@@ -0,0 +1,98 @@
+            }
+        } catch (SQLException | JsonProcessingException e) {
+            String message = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
+            log.warn("Failed to summarize '{}': {}", pref, message);
```
Is it ok to just log a warning and return an empty List?

hackerdave replied:
The interface doesn't provide for an exception

    public List<TextSegment> split(Document document) {

Perhaps we should throw a runtime exception?

        try {
          ...
        } catch (SQLException | JsonProcessingException e) {
          throw new RuntimeException(e);
        }

psilberk replied:
Yes, particularly if there's no practical reason on trying to recover from this condition (not sure if it does, just stating the reasoning). And yes for a RuntimeException or subclass.. You could also add a message with the original exception. Another idea would be to check other implementations in case the library already has a RuntimeException subclass we can use.

hackerdave replied:
Changed to throw a RuntimeException

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/model/oracle/OracleEmbeddingModel.java:
```diff
@@ -0,0 +1,169 @@
+            }
+        } catch (SQLException | JsonProcessingException e) {
+            String message = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
+            log.warn("Failed to summarize '{}': {}", pref, message);
```
Similar to the other comment, is it ok to warn the exception and return what embeddings got to that moment?

hackerdave committed:
Throw runtime exception, add loader pref class

hackerdave committed:
Add javadoc

hackerdave committed:
Add javadoc

hackerdave committed:
Add javadoc

hackerdave committed:
Rename getDir

hackerdave committed:
Make constant

hackerdave committed:
Change to separate file, dir, and table pref

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/DirectoryPreference.java:
```diff
@@ -0,0 +1,31 @@
+ * {"dir": "directory name"}
+ */
+public class DirectoryPreference {
+    private String dir;
```
I think you don't need the @JsonProperty annotation if you just use 'directory' and standard setters/getters

hackerdave replied:
Jackson apparently uses the getter/setter name rather than the variable name to map to the JSON so we do need it when the getter/setter doesn't match the JSON such as in this case where the getter/setter uses getDirectory/setDirectory but the JSON is {"dir": "..."}. I thought it would be better to be explicit.

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/FilePreference.java:
```diff
@@ -0,0 +1,31 @@
+ * {"file": "filename"}
+ */
+public class FilePreference {
+    private String file;
```
Would it make more sense to call the property 'filename'?

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -51,17 +52,20 @@ public List<Document> loadDocuments(String pref) throws JsonProcessingException,
-
-        if (loaderPref.getFile() != null && !loaderPref.getFile().equals("null")) {
-            String filename = loaderPref.getFile();
+        // disable throwing an exception on properties not listed in the class
```
Can we add a sample of the JSON in the javadoc?

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -76,22 +80,13 @@ public List<Document> loadDocuments(String pref) throws JsonProcessingException,
-                throw new InvalidParameterException("Missing owner in preference");
-            }
-
+        } else if (tablePref.isValid()) {
```
Maybe we can add 'path' in the RuntimeException

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/OracleDocumentLoader.java:
```diff
@@ -76,22 +80,13 @@ public List<Document> loadDocuments(String pref) throws JsonProcessingException,
             documents.addAll(loadDocuments(owner, table, column, pref));
         } else {
-            throw new InvalidParameterException("Missing file, dir, or table in preference");
+            throw new InvalidParameterException("Invalid preference");
```
I think we should add more information in here.

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/TablePreference.java:
```diff
@@ -1,35 +1,24 @@
- * for a directory: {"dir": "directory name"}
- * for a table:     {"owner": "owner", "tablename": "table name", "colname": "column name"}
+ * To specify a table, dbms_vector_chain.utl_to_text expects the following JSON:
+ * {"owner": "owner", "tablename": "table name", "colname": "column name"}
```
"colname" is the name of the field?

hackerdave replied:
Unfortunately yes so we won't be able to change the JSON schema

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/data/document/loader/oracle/TablePreference.java:
```diff
@@ -1,35 +1,24 @@
-
-    private String file;
-    private String dir;
+public class TablePreference {
```
Im not sure if this is the best name for the class. But maybe because it's not documented what it represents. What does this represent?

hackerdave replied:
This is to specify a table and column to load documents from rather than from the file system. It's intended to be used by OracleDocumentLoader to parse the JSON preference specified by a user but a user could also use this to build the JSON preference. Similarly FilePreference and DirectoryPreference for loading files from the file system.

hackerdave committed:
Rename property

hackerdave committed:
More javadoc

hackerdave committed:
Better error messages

hackerdave committed:
Better error messages

hackerdave committed:
Remove slf4j to match other tests

hackerdave committed:
Functions should be public

psilberk commented on langchain4j-oracle/src/main/java/dev/langchain4j/model/oracle/OracleEmbeddingModel.java:
```diff
@@ -0,0 +1,206 @@
+
+        if (!batching) {
+            for (String input : inputs) {
+                String query = "select t.column_value as data from dbms_vector_chain.utl_to_embeddings(?, json(?)) t";
```
Are lines 135 to 160 exactly the same to 158 to 173? (Except for the input value for 1st object in stmt). Maybe we can create a common method for this?

hackerdave replied:
Yes they are the same. Refactored to use a common method and added a test for batching. It's actually for non-batching since the default is batching.

hackerdave committed:
Use common method for embedding

hackerdave committed:
Add batching test

psilberk commented on langchain4j-oracle/pom.xml:
```diff
@@ -112,6 +113,28 @@
         </dependency>
 
+        <!-- document store dependencies -->
+        <dependency>
```
Lets change the order so the one for test are still at the bottom.
Can you check if they are not already in the langchain4j-core?
Otherwise we need approval for those dependencies that are not for test.

hackerdave replied:
Changed the order so they are listed before test. Also changed jackson to use the parent since langchain4j-parent is including it.

hackerdave committed:
Move above test dependencies

hackerdave committed:
Change jsoup version

hackerdave committed:
Change input to clob

hackerdave committed:
Add table metadata

hackerdave committed:
Remove comment
